### PR TITLE
Support report accumulation for `casr-libfuzzer` and `casr-afl`

### PR DESCRIPTION
**Pull request by hkctkuy**

### CODE REVIEW

**Review by SweetVishnya — CHANGES_REQUESTED** [review]

File: casr/tests/tests.rs [hunk 1]
```diff
@@ -3951,32 +3951,32 @@ fn test_casr_libfuzzer() {
...
+    let output = cmd.output().expect("failed to start casr-libfuzzer");
     assert!(
         output.status.success(),
         "Stdout {}.\n Stderr: {}",
         String::from_utf8_lossy(&output.stdout),
         String::from_utf8_lossy(&output.stderr)
     );
-    let err = String::from_utf8_lossy(&output.stderr);
 
+    let err = String::from_utf8_lossy(&output.stderr);
```
[the remark this answers is not in the document]
Comment (hkctkuy) [reply casr/tests/tests.rs:3979]: Nothing. Idk, why git highlighted it. I Just add empty line before it.

File: casr/src/util.rs [hunk 2]
```diff
@@ -316,28 +317,36 @@ pub fn get_atheris_lib() -> Result<String> {
...
-            fs::create_dir_all(output_dir).with_context(|| {
-                format!("Couldn't create output directory {}", output_dir.display())
-            })?;
         } else {
             bail!("Output directory is not empty.");
         }
     }
+
+    if let Some(seed_dir) = matches.get_one::<PathBuf>("seed") {
+        copy_dir(seed_dir, output_dir)
```
[the remark this answers is not in the document]
Comment (hkctkuy) [reply casr/src/util.rs:329]: The output directory cannot be empty. A non-empty directory results in an error or directory re-creation if the `--force-remove` flag is present.

[the remark this answers is not in the document]
Comment (hkctkuy) [reply casr/src/util.rs:329]: See code above

File: casr/src/util.rs [hunk 3]
```diff
@@ -316,28 +317,36 @@ pub fn get_atheris_lib() -> Result<String> {
...
+            .with_context(|| format!("Couldn't copy seed directory {}", seed_dir.display()))?;
+        // Get casrep dir
+        let casrep_dir = output_dir.join("casrep");
+        if !casrep_dir.exists() && fs::create_dir_all(&casrep_dir).is_err() {
+            bail!("Failed to create dir {}", &casrep_dir.to_str().unwrap());
+        }
+    } else {
+        fs::create_dir_all(output_dir).with_context(|| {
```
[the remark this answers is not in the document]
Comment (hkctkuy) [reply casr/src/util.rs:337]: It is checked above

[the remark this answers is not in the document]
Comment (hkctkuy) [reply casr/src/util.rs:337]: Such corner case does not result in an error

[the remark this answers is not in the document]
Comment (hkctkuy) [reply casr/src/util.rs]: Patched

File: casr/src/bin/casr-afl.rs [hunk 4]
```diff
@@ -75,6 +75,16 @@ fn main() -> Result<()> {
                 .value_parser(clap::value_parser!(PathBuf))
                 .help("Output directory with triaged reports")
         )
+        .arg(
```
Comment (SweetVishnya) [inline casr/src/bin/casr-afl.rs:78]: docs/usage.md should be updated

Comment (hkctkuy) [reply casr/src/bin/casr-afl.rs:78]: Done

File: casr/src/triage.rs [hunk 5]
```diff
@@ -230,41 +238,66 @@ pub fn fuzzing_crash_triage_pipeline(
...
+                .output()
+                .with_context(|| format!("Couldn't launch {casr_cluster:?}"))?;
+
+            if casr_cluster_c.status.success() {
+                info!(
+                    "{}",
+                    String::from_utf8_lossy(&casr_cluster_c.stdout).trim_end()
+                );
+            } else {
+                error!(
```
Comment (SweetVishnya) [inline casr/src/triage.rs:289]: why bail changed to error here?

**Review by SweetVishnya — APPROVED** [review]

File: casr/src/bin/casr-afl.rs [hunk 6]
```diff
@@ -75,6 +75,16 @@ fn main() -> Result<()> {
...
+            Arg::new("seed")
```
Comment (anfedotoff) [inline casr/src/bin/casr-afl.rs:79]: I think, we need to find an appropriate name for this option.
Seed is confusing.

In fuzzing  seed is input data that program processes normally.
Also, seed is used for random generators.

Please remove, short option too, only long is needed.

**Review by anfedotoff — CHANGES_REQUESTED** [review]

File: casr/src/bin/casr-afl.rs (hunk 6 above)
Comment (SweetVishnya) [reply casr/src/bin/casr-afl.rs:79]: Maybe, this way?

--update-clusters PREV_CRASHES_DIR

Update clusters in PREV_CRASHES_DIR with new crashes from INPUT_DIR and store accumulated crash report clusters to OUTPUT_DIR

CASR_PREV_CRASHES_DIR

File: casr/Cargo.toml [hunk 7]
```diff
@@ -37,6 +37,7 @@ tokio = { version = "1", features = ["rt", "macros"], optional = true }
 toml = { version = "0.8", optional = true }
 wait-timeout = "0.2"
 which = "6.0"
+copy_dir = "0.1.3"
```
[the remark this answers is not in the document]
Comment (hkctkuy) [reply casr/Cargo.toml]: Done

File: casr/src/bin/casr-afl.rs (hunk 6 above)
Comment (hkctkuy) [reply casr/src/bin/casr-afl.rs:79]: Why only the long version?

Comment (anfedotoff) [reply casr/src/bin/casr-afl.rs:79]: > Maybe, this way?
> 
> --update-clusters PREV_CRASHES_DIR
> 
> Update clusters in PREV_CRASHES_DIR with new crashes from INPUT_DIR and store accumulated crash report clusters to OUTPUT_DIR
> 
> CASR_PREV_CRASHES_DIR

LGTM

Comment (hkctkuy) [reply casr/src/bin/casr-afl.rs]: Renamed

Comment (hkctkuy) [reply casr/src/bin/casr-afl.rs]: Renamed x2

File: casr/src/bin/casr-afl.rs [hunk 8]
```diff
@@ -76,14 +76,13 @@ fn main() -> Result<()> {
                 .help("Output directory with triaged reports")
         )
         .arg(
-            Arg::new("base")
-                .short('b')
-                .long("base")
-                .env("CASR_BASE_DIR")
+            Arg::new("join")
+                .long("join")
+                .env("CASR_PREV_CLSUTERS_DIR")
```
[the remark this answers is not in the document]
Comment (SweetVishnya) [reply casr/src/bin/casr-afl.rs:81]: and other similar places

**Review by anfedotoff — APPROVED** [review]

File: docs/usage.md [hunk 9]
```diff
@@ -447,19 +449,22 @@ Triage crashes found by AFL++/Sharpfuzz
...
-      -V, --version                Print version
+      -l, --log-level <log-level>     Logging level [default: info] [possible values: info,
+                                      debug]
+      -j, --jobs <jobs>               Number of parallel jobs for generating CASR reports
+                                      [default: half of cpu cores]
+      -t, --timeout <SECONDS>         Timeout (in seconds) for target execution, 0 value means
+                                      that timeout is disabled [default: 0]
+      -i, --input <INPUT_DIR>         AFL++ work directory
+      -o, --output <OUTPUT_DIR>       Output directory with triaged reports
+          --join <PREV_CLSUTERS_DIR>  Use directory with previously triaged reports for new
```
Comment (SweetVishnya) [inline docs/usage.md:460]: fix typos

**Review by SweetVishnya — CHANGES_REQUESTED** [review]

File: docs/usage.md (hunk 9 above)
Comment (hkctkuy) [reply docs/usage.md]: Done

File: casr/src/bin/casr-afl.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (hkctkuy) [reply casr/src/bin/casr-afl.rs]: Done

**Review by SweetVishnya — APPROVED** [review]

**Pull request merged**